{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This solution the methylation calls file from bismark as the input instead of the bedgraph file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count the number of methylated/non-methylated sites for each CpG site\n",
    "SRR3083026_CpG_file = \"/Users/cmdb/qbb2020-answers/assignment6/CpG_context_SRR3083926_1.chr6_bismark_bt2_pe.txt\"\n",
    "SRR3083029_CpG_file = \"/Users/cmdb/qbb2020-answers/assignment6/CpG_context_SRR3083929_1.chr6_bismark_bt2_pe.txt\"\n",
    "\n",
    "#bismark_methylation_extractor format\n",
    "'''\n",
    "(1) seq-ID\n",
    "(2) methylation state, + for methylated, - for unmethylated\n",
    "(3) chromosome\n",
    "(4) start position (= end position)\n",
    "(5) methylation call\n",
    "'''\n",
    "\n",
    "#save methyl calls where site bp in chr6 is the key, with elements as two element list\n",
    "#first element of list is # of methyl calls, second is # of unmethylated calls\n",
    "SRR3083026_CpG_calls = {}\n",
    "\n",
    "SRR3083029_CpG_calls = {}\n",
    "\n",
    "#parse SRR3083026_CpG_file \n",
    "with open(SRR3083026_CpG_file, 'r') as f:\n",
    "    for line in f:\n",
    "        #skip first line header\n",
    "        if \"Bismark\" in line:\n",
    "            continue\n",
    "        fields = line.rstrip().split('\\t')\n",
    "        methyl_call = fields[1]\n",
    "        site = fields [3]\n",
    "        if site not in SRR3083026_CpG_calls:\n",
    "            SRR3083026_CpG_calls[site] = [0,0]\n",
    "        if methyl_call == \"+\":\n",
    "            SRR3083026_CpG_calls[site][0] +=1\n",
    "        else: \n",
    "            SRR3083026_CpG_calls[site][1] +=1        \n",
    "        \n",
    "\n",
    "#parse SRR3083029_CpG_file\n",
    "with open(SRR3083029_CpG_file, 'r') as f:\n",
    "    for line in f:\n",
    "        #skip first line header\n",
    "        if \"Bismark\" in line:\n",
    "            continue\n",
    "        fields = line.rstrip().split('\\t')\n",
    "        methyl_call = fields[1]\n",
    "        site = fields [3]\n",
    "        if site not in SRR3083029_CpG_calls:\n",
    "            SRR3083029_CpG_calls[site] = [0,0]\n",
    "        if methyl_call == \"+\":\n",
    "            SRR3083029_CpG_calls[site][0] +=1\n",
    "        else: \n",
    "            SRR3083029_CpG_calls[site][1] +=1  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter out redundant gene coordinates from chr6 bed file\n",
    "#5th and 6th columns are gene coordinates while the 13th column is the gene\n",
    "\n",
    "input_bedfile = \"/Users/cmdb/qbb2020-answers/assignment6/mm10_refseq_genes_chr6_50M_60M.bed\"\n",
    "filtered_bedfile = \"/Users/cmdb/qbb2020-answers/assignment6/mm10_refseq_genes_chr6_50M_60M_filtered.bed\"\n",
    "\n",
    "#keep track of unique genes detected\n",
    "#use dictionary as a hash table for look-up instead of a list which will be slower\n",
    "\n",
    "genes = {}\n",
    "output = open(filtered_bedfile, 'w')\n",
    "with open(input_bedfile, 'r') as f:\n",
    "    for line in f: \n",
    "        fields = line.rstrip().split('\\t')\n",
    "        gene = fields[12]\n",
    "        gene_start = fields[4]\n",
    "        gene_end = fields[5]\n",
    "        if gene not in genes:\n",
    "            genes[gene] = [gene_start, gene_end]\n",
    "            output.write(line)\n",
    "        \n",
    "output.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert dictionary to list of tuples (to allow sorting), where first element of tuple is the gene name\n",
    "#second element is gene start location, and the third element is gene end location\n",
    "gene_loci = []\n",
    "gene_keys = genes.keys()\n",
    "for key in gene_keys:\n",
    "    gene_location = genes[key]\n",
    "    gene_locus = (key, gene_location[0], gene_location[1])\n",
    "    gene_loci.append(gene_locus)\n",
    "\n",
    "\n",
    "#sort the gene list. This should technically be the same as the gene_loci list since the original bed file looked sorted\n",
    "#this is just in-case the gene bed file was not sorted\n",
    "sorted_gene_loci = sorted(gene_loci, key=lambda locus: locus[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save methyl calls where site bp in genes are the keys, with elements as two element list\n",
    "#first element of list is # of methyl calls, second is # of unmethylated calls\n",
    "SRR3083026_gene_methylcounts = {}\n",
    "for loci in sorted_gene_loci:\n",
    "    gene_name = loci[0]\n",
    "    SRR3083026_gene_methylcounts[gene_name] = [0,0]\n",
    "\n",
    "\n",
    "methyl_positions = SRR3083026_CpG_calls.keys()\n",
    "for position in methyl_positions:\n",
    "\n",
    "    for loci in sorted_gene_loci:\n",
    "        gene = loci[0]\n",
    "        start_bp = int(loci[1])\n",
    "        end_bp = int(loci[2])\n",
    "        if int(position) >= start_bp and int(position)<=end_bp:\n",
    "            methyl_calls = SRR3083026_CpG_calls[position]\n",
    "            methylated = methyl_calls[0]\n",
    "            unmethylated = methyl_calls[1]\n",
    "            SRR3083026_gene_methylcounts[gene][0] += methylated\n",
    "            SRR3083026_gene_methylcounts[gene][1] += unmethylated\n",
    "            break\n",
    "\n",
    "#do the same for SRR3083029\n",
    "SRR3083029_gene_methylcounts = {}\n",
    "for loci in sorted_gene_loci:\n",
    "    gene_name = loci[0]\n",
    "    SRR3083029_gene_methylcounts[gene_name] = [0,0]\n",
    "\n",
    "\n",
    "methyl_positions = SRR3083029_CpG_calls.keys()\n",
    "for position in methyl_positions:\n",
    "\n",
    "    for loci in sorted_gene_loci:\n",
    "        gene = loci[0]\n",
    "        start_bp = int(loci[1])\n",
    "        end_bp = int(loci[2])\n",
    "        if int(position) >= start_bp and int(position)<=end_bp:\n",
    "            methyl_calls = SRR3083029_CpG_calls[position]\n",
    "            methylated = methyl_calls[0]\n",
    "            unmethylated = methyl_calls[1]\n",
    "            SRR3083029_gene_methylcounts[gene][0] += methylated\n",
    "            SRR3083029_gene_methylcounts[gene][1] += unmethylated\n",
    "            break\n",
    "\n",
    "#print(SRR3083026_gene_methylcounts)           \n",
    "#print(SRR3083029_gene_methylcounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Mpp6', 'Dfna5', 'Osbpl3', 'Cycs', '5430402O13Rik', '4921507P07Rik', 'Npvf', 'C530044C16Rik', 'Mir148a', 'Gm6559', 'Nfe2l3', 'Hnrnpa2b1', 'Cbx3', 'Snx10', 'Skap2', 'Halr1', 'Hotairm1', 'Hoxa2', 'Hoxaas2', 'Hoxa3', 'Hoxa7', 'Hoxa9', 'Mir196b', 'Hoxa10', 'Hoxa11', 'Hoxa11os', 'Hoxa13', 'Hottip', 'Evx1os', 'Evx1', '1700094M24Rik', 'Hibadh', 'Tax1bp1', 'Jazf1', 'Gm4872', '9430076C15Rik', 'Creb5', 'Tril', 'Cpvl', '4921529L05Rik', 'Chn2', '9130019P16Rik', 'Wipf3', 'Scrn1', 'Fkbp14', 'Plekha8', 'Mturn', 'Znrf2', 'Nod1', 'Ggct', 'Gars', 'Crhr2', 'Inmt', 'Fam188b', 'Aqp1', 'Ghrhr', '6430584L05Rik', 'Adcyap1r1', 'Neurod6', 'Gm3279', 'Ccdc129', 'Ppp1r17', 'Pde1c', 'Lsm5', 'Avl9', 'Kbtbd2', 'Fkbp9', 'Nt5c3', 'Vmn1r4', 'Vmn1r5', 'Vmn1r7', 'Vmn1r8', 'Vmn1r11', 'Vmn1r13', 'Vmn1r14', 'Vmn1r15', 'Ppm1k', 'Herc6', 'Pyurf', 'Lancl2', 'Vopp1', 'Vmn1r27', 'Vmn1r29', 'Vmn1r30', 'Abcg2', 'Herc3', 'Fam13a', 'Tigd2', 'Gprin3']\n",
      "[2.849144510876479, 4.31040997543926, 4.426348831131807, 7.212121212121212, 2.861019577437488, 3.9000000000000004, 3.34855403348554, 2.1199055538445744, 0.0, 2.1923859343214183, 4.148133470714115, 3.470052083333333, 5.925311203319503, 5.070666443215463, 2.7, 16.538551401869157, 6.838709677419355, 2.9647058823529413, 2.9690721649484537, 2.034792731035375, 1.0982142857142856, 3.5710843373493977, 0.0, 0.8096249115357396, 0.7359550561797753, 5.435643564356435, 1.8199999999999998, 1.5161744022503516, 1.8782771535580522, 0.2181208053691275, 4.1063174114021574, 2.7701148251062437, 2.772766058778005, 2.643251192092165, 22.28125, 4.968851063829788, 4.532062410426976, 2.331081081081081, 4.112555826898114, 4.507246376811595, 2.6999250782078357, 2.7175548087694033, 6.03397341211226, 6.902019546887931, 2.9467181467181467, 3.1908289730071906, 8.654718137254903, 4.927272727272728, 5.338764823656246, 4.471631205673759, 2.724544170518747, 5.120199146514936, 5.133333333333333, 2.456612685560054, 7.067215363511659, 5.75098814229249, 15.66536964980545, 2.4053843759726115, 5.230769230769231, 6.291128337639965, 3.724522011961957, 13.74841972187105, 2.440365905227387, 2.845528455284553, 2.4264039408866993, 5.612735995047973, 2.9323589032326893, 3.0898581187598526, 0.9000000000000001, 1.5, 0.8333333333333334, 6.176470588235294, 1.0, 0.8571428571428571, 0.8095238095238095, 5.000000000000001, 2.8493575207860924, 3.3592110685201164, 1.5320427236315086, 2.6431393052166845, 4.441924528301887, 6.0, 1.2, 4.0, 3.4332816508555584, 2.2256633447478986, 4.53603335318642, 2.7662835249042144, 5.6468174136959846]\n"
     ]
    }
   ],
   "source": [
    "genes = SRR3083026_gene_methylcounts.keys()\n",
    "methylated_genes = []\n",
    "\n",
    "fold_changes = []\n",
    "\n",
    "for gene in genes:\n",
    "    #calculate percentage methylation for SRR3083026\n",
    "    SRR3083026_counts = SRR3083026_gene_methylcounts[gene]\n",
    "    methyl_counts = SRR3083026_counts[0]\n",
    "    unmethyl_counts = SRR3083026_counts[1]\n",
    "    #skip gene if methylation for E4 cell is zero\n",
    "    if methyl_counts == 0:\n",
    "        continue\n",
    "    total_counts = methyl_counts + unmethyl_counts \n",
    "    #skip gene since no count data at all for gene\n",
    "    if total_counts == 0:\n",
    "        continue\n",
    "    SRR3083026_pct_methyl = methyl_counts/total_counts\n",
    "    \n",
    "    #calculate percentage methylation for SRR3083029\n",
    "    SRR3083029_counts = SRR3083029_gene_methylcounts[gene]\n",
    "    methyl_counts = SRR3083029_counts[0]\n",
    "    unmethyl_counts = SRR3083029_counts[1]\n",
    "    total_counts = methyl_counts + unmethyl_counts \n",
    "    #skip gene since no count data at all for gene\n",
    "    if total_counts == 0:\n",
    "        continue\n",
    "    SRR3083029_pct_methyl = methyl_counts/total_counts\n",
    "    \n",
    "    #calculate fold change\n",
    "    fold_change = SRR3083029_pct_methyl/SRR3083026_pct_methyl\n",
    "    methylated_genes.append(gene)\n",
    "    fold_changes.append(fold_change)\n",
    "\n",
    "        \n",
    "print(methylated_genes)\n",
    "print(fold_changes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file = \"/Users/cmdb/qbb2020-answers/assignment6/fold-methylation.txt\"\n",
    "with open(output_file, 'w') as f:\n",
    "    f.write(\"gene\" + '\\t' + 'fold_change, E5.5/E4.0' + '\\n')\n",
    "    for i, gene in enumerate(methylated_genes):\n",
    "        f.write(gene + '\\t' + str(fold_changes[i]) + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
